src/history/hs_rec.c

```diff
@@ -207,54 +110,53 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
-         *
-         * Verify simple checks first to confirm whether the retrieved update same or not before
-         * performing the expensive key comparison.
-         */
-        if (hs_btree_id == btree->id && tw->start_ts == hs_start_ts) {
-            WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
 
 #ifdef HAVE_DIAGNOSTIC
-            if (cmp == 0) {
-                WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
-                  &upd_type_full_diag, existing_val));
-                WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
-                /*
-                 * Check if the existing HS value is same as the new value we are about to insert.
-                 * We can skip this check if the existing value has a globally visible stop time,
-                 * i.e., the value has been deleted from the HS.
-                 */
-                if (cmp == 0)
-                    WT_ASSERT(session,
-                      (WT_TIME_WINDOW_HAS_STOP(&hs_cbt->upd_value->tw) &&
-                        __wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw)) ||
-                        tw->start_txn == WT_TXN_NONE ||
-                        tw->start_txn != hs_cbt->upd_value->tw.start_txn ||
-                        tw->start_ts != hs_cbt->upd_value->tw.start_ts);
-                counter = hs_counter + 1;
-            }
-#else
+        if (tw->start_ts == hs_start_ts) {
+            WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
+              &upd_type_full_diag, existing_val));
+            WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
+            /*
+             * We shouldn't be inserting the same value again for the key unless coming from a
+             * different transaction. If the updates are from the same transaction, the start
+             * timestamp for each update should be different.
+             */
             if (cmp == 0)
-                counter = hs_counter + 1;
-#endif
+                WT_ASSERT(session,
+                  tw->start_txn == WT_TXN_NONE ||
+                    tw->start_txn != hs_cbt->upd_value->tw.start_txn ||
+                    tw->start_ts != hs_cbt->upd_value->tw.start_ts);
+            counter = hs_counter + 1;
         }
+#else
+        if (tw->start_ts == hs_start_ts)
+            counter = hs_counter + 1;
+#endif
     }
 
     /*
      * If we're inserting a non-zero timestamp, look ahead for any higher timestamps. If we find
      * updates, we should remove them and reinsert them at the current timestamp.
      */
     if (tw->start_ts != WT_TS_NONE) {
-        WT_ERR_NOTFOUND_OK(__wt_hs_cursor_next(session, cursor), true);
+        /*
+         * If there were no keys equal to or less than our target key, we would have received
+         * WT_NOT_FOUND. In that case we need to search again with a higher timestamp.
```

Comment:
WT_NOTFOUND?

Reply:
Add more details to the comments why we are again doing the search.

Reply:
Fixed.